src/main/scala/metadata.scala

```diff
@@ -0,0 +1,83 @@
+package suffuse
+
+/** AKey is an attribute key: a Metadata map key which carries
+ *  both the type of the corresponding value, and a value to
+ *  provide when there is no value in the map. This as opposed
+ *  to wrapping everything in sight in Option.
+ */
+final class AKey[A](description: String, empty: => A) {
+  def emptyValue: A = empty
+  override def toString = description
+}
+
+/** Attr is a dependently typed key/value pair.
+ */
+trait Attr extends Any {
+  type Type
+  def key: AKey[Type]
+  def value: Type
+
+  def pair                               = key -> value
+  def matches[A](implicit that: AKey[A]) = key == that
+  def to_s                               = s"$key: $value"
+}
+
+/** The simple implementation of Attr, which ties the knot between the
+ *  user-facing type parameter and the type member which couples key and value.
+ */
+final case class AttrOf[A](key: AKey[A], value: A) extends Attr {
+  type Type = A
+  override def toString = s"$key: $value"
+}
+
+object Attr {
+  def apply[A](value: A)(implicit key: AKey[A]): AttrOf[A] = new AttrOf[A](key, value)
+  def unapply(x: Attr): Some[(AKey[x.Type], x.Type)]       = Some(x.pair)
+}
+
+/** A Metadata map, holding any number of well typed Attrs.
+ *  A typed value can be obtained for any key.
+ */
+final class Metadata(attrs: Vector[Attr]) {
+  private[this] val untypedMap                   = attrs.foldLeft(Map[AKey[_], Any]())(_ + _.pair)
+  private def doApply[A]()(implicit z: AKey[A]): A = untypedMap(z).asInstanceOf[A]
+
+  def isEmpty                                = untypedMap.isEmpty
+  def apply[A]()(implicit z: AKey[A]): A     = if (has[A]) doApply[A]() else z.emptyValue
+  def has[A]()(implicit z: AKey[A]): Boolean = keys contains z
+  def keys: Vector[AKey[_]]                  = attrs map (_.key)
+
+  /** Set could also hold onto the old value, and just return the last one.
+   *  It would amount to preserving the attribute's history.
+   */
+  def set(attr: Attr): Metadata        = new Metadata(attrs.filterNot(_.key eq attr.key) :+ attr)
+  def set[A: AKey](value: A): Metadata = set(Attr[A](value))
+
+  override def toString = if (isEmpty) "{ }" else attrs mkString ("{\n  ", "\n  ", "\n}")
+}
+object Metadata {
+  private[this] val Empty: Metadata = new Metadata(Vector())
+
+  def empty(): Metadata             = Empty
+  def apply(attrs: Attr*): Metadata = attrs.foldLeft(empty)(_ set _)
+}
+
+object Example {
+  /** Implicit keys are the road to strongly typed and usable both.
+   */
+  implicit val mtime = new AKey[Mtime]("modification time", Mtime(0))
+  implicit val atime = new AKey[Atime]("access time", Atime(0))
+  implicit val size  = new AKey[Size]("size in bytes", Size(-1L))
+
+  final case class Mtime(val timestamp: Long)
```

Comment:
> some sick tradeoff scala will otherwise inflict upon ye

Yes, and with these things it seems one of our key differences of opinion come in. I tend to lean towards usability and trade in security. You go from the other direction. In this example, ...[removed text]..., we both know how we would want this.

Maybe we can compromise and clearly state our intentions while having (if possible) a friendly user experience:

```
/* final */ class AKey {
  ...
}

/*
  type mtime { timestamp: Long }
  def mtime(timestamp: Long): mtime
  val mtime: AKey[mtime]
*/
final case class mtime(timestamp: Long)
object mtime extends AKey[mtime]("description") 
``` 

While the above construction allows us to have a term and type that have the same name. We compromise the following:
1. `AKey` can be extended by the end user
2. `object mtime` introduces a new type

The problem with `1.` is that it would allow the user to use the keyword (which probably hate most) `override`
The problem with `2.` is that it can cause funny things with type related code

I wonder if we should go with the end user version and become more safe as soon as we hit a problem.